Prevent credentials edition for sites with no purchases

monsieur-z opened the pull request:
### Changes proposed in this Pull Request

Sites that never had any purchased plan or product can't currently have user credentials be saved. See 1164141197617539-as-1199981665376903. This PR improves the user experience a bit by showing an empty state if the site has no purchased plans nor products.

### Implementation notes

#### Changing site

When making a request to fetch a site purchases, information about the state of the request (ongoing and completed) is stored in the state independently of the site (respectively `state.purchases.isFetchingSitePurchases` and `state.purchases.hasLoadedSitePurchasesFromServer`). This is limiting in our case, since the loading condition in `HasSitePurchasesSwitch` will keep returning false after switching site.

To prevent this, we could isolate the site specific request information in the state (e.g. `state.purchases.requests[siteId].isFetching`), but that's too big of a change, at least for the scope of this PR.

A simpler solution is to detect the site switch in `HasSitePurchasesSwitch`, and to reset the `isFetchingSitePurchases` and `hasLoadedSitePurchasesFromServer` values.

#### Testing

Unit tests were added for the newly created components. Most of them doesn't include any logic, so I've just included snapshots to make sure further changes are intentional.

### Testing instructions

- Download the PR and run cloud
- Create a new self-hosted Jetpack site, and connect it (don't purchase any plan or product)
- Visit `/settings/:site`
- Notice the loading placeholder while purchases are being fetched, then the empty state (see video below)
- Check that the button links to `jetpack.com/pricing`
- Switch section back and forth, and check that you see the empty state instantly
- Select another site that has a purchased product or plan
- Check that you can add your credentials
- Remove any purchased product or plan
- Make sure you can still remove or edit the credentials

### Screenshots


https://user-images.githubusercontent.com/1620183/111787478-5905f780-8895-11eb-833c-9448796ed3ea.mov

Commit:
Prevent credentials edition for sites with no purchases

Commit:
Allow credentials removal and edition if creds set

monsieur-z commented:
Any feedback for the notice copy is welcome.

rcanepa reviewed (approved):
Code changes look good and the form behaves as expected. 

LGTM!

Commit:
Add empty state in setting section when no purchase

rcanepa commented:
@monsieur-z, let me know if you need a second review.

Commit:
Add tests to AdvancedCredentialsLoadingPlaceholder

Commit:
Simplify CSS rules in empty settings page

Commit:
Add tests to NoSitePurchasesMessage

Commit:
Add tests to HasSitePurchasesSwitch

Commit:
Update loading placeholder styles

Commit:
Force credentials section update when switching site

monsieur-z commented on client/components/has-site-purchases-switch/index.tsx:
```diff
@@ -0,0 +1,63 @@
+	useEffect( () => {
+		if ( siteId !== currentSiteId ) {
+			setCurrentSiteId( siteId );
+			dispatch( resetSiteState() );
```
Will reset loading information. See implementation notes.

monsieur-z commented on client/components/has-site-purchases-switch/test/index.jsx:
```diff
@@ -0,0 +1,82 @@
+import HasSitePurchasesSwitch from '../index';
+import { reducer as purchases } from 'calypso/state/purchases/reducer';
+
+jest.mock( 'calypso/components/data/query-site-purchases', () => () => <p>Query</p> );
```
It was easier here to mock the query component rather than mocking the endpoint.

monsieur-z commented on client/jetpack-cloud/sections/settings/controller.tsx:
```diff
@@ -18,7 +22,17 @@ export const settings: PageJS.Callback = ( context, next ) => {
-	context.primary = <AdvancedCredentials action={ action } host={ host } role="main" />;
+	const siteId = getSelectedSiteId( context.store.getState() ) as number;
+
+	context.primary = (
```
Site purchases now need to be fetched before rendering anything.

monsieur-z commented on client/jetpack-cloud/sections/settings/loading/style.scss:
```diff
@@ -0,0 +1,12 @@
+		@include placeholder();
+
+		width: 100%;
+		height: 186px;
```
Prevent illustration from "jumping" after it has finished loading.

monsieur-z commented:
> @monsieur-z, let me know if you need a second review.

Thanks @rcanepa. I've updated the instructions, since changes are quite significative.

mattgawarecki commented:
*(EDIT: See my next comment)* ~Everything~ Almost everything you mention in the PR description works for me, and the unit tests are clear. I know snapshots aren't ideal, but given there's no logic, I think this is much better than not having them at all.

I was _just_ about to approve, but I noticed on mobile layouts there's no page title in the top bar and no way for me to switch sites (see screenshots). Can you add those in?

**With purchases (left) / without purchases (right)**

<img height="300" alt="image" src="https://user-images.githubusercontent.com/670067/111837432-dcd7d800-88c5-11eb-9c9e-47bcd5a1ab32.png"> <img height="300" alt="image" src="https://user-images.githubusercontent.com/670067/111837698-42c45f80-88c6-11eb-82c1-988e1f6d50eb.png">

mattgawarecki commented:
Actually, I just went back to confirm I could delete credentials from a site with no purchases but with credentials having been saved beforehand. It looks like maybe not? I'm getting the "empty content" page when I try.

monsieur-z commented:
> I was _just_ about to approve, but I noticed on mobile layouts there's no page title in the top bar and no way for me to switch sites (see screenshots). Can you add those in?

I'm not seeing the Manage Sites dropdown at any viewport, @mattgawarecki. Do you know what are the conditions to show it?

Commit:
Don't show empty settings state for sites with credentials

monsieur-z commented:
> Actually, I just went back to confirm I could delete credentials from a site with no purchases but with credentials having been saved beforehand. It looks like maybe not? I'm getting the "empty content" page when I try.

You're right, I forgot to carry over that part after I changed the implementation. Updated.

monsieur-z commented on client/jetpack-cloud/sections/settings/controller.tsx:
```diff
@@ -18,7 +23,25 @@ export const settings: PageJS.Callback = ( context, next ) => {
+	const sectionElt = <AdvancedCredentials action={ action } host={ host } role="main" />;
+
+	context.primary = (
+		<HasSiteCredentialsSwitch
```
The site credentials and purchases requests should ideally be done concurrently, but I favoured having 2 separate switches here for clarity. It seems that credentials data are often already in the state at this point, so the gain of doing that seems to be minimal.

monsieur-z commented on client/jetpack-cloud/sections/settings/has-site-credentials-switch.tsx:
```diff
@@ -0,0 +1,70 @@
+	const hasLoaded = siteStatus === 'success';
+	const hasCredentials = Object.keys( credentials ).length > 0;
+
+	const [ isLocked, setLocked ] = useState( hasCredentials );
```
We want to prevent further calls to the credentials endpoint to trigger the loading state again.

mattgawarecki reviewed (approved):
Everything tested well for me this time. Nice work! 👍 

Sidenote: I think we'll eventually want to come back and make some copy and/or design changes to reflect this "credentials exist but no active purchases" state, since we still mention Jetpack being able to perform restores and scan for threats.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/5647767

Hi @monsieur-z, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include all of the following strings:
- `Upgrade for quick restores`
- `Server credentials are used to restore your site and fix any vulnerabilities`

Thank you in advance!

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/5647767).
